### PR TITLE
[FIX] website_sale: limit pricelist to it's website if set

### PR DESCRIPTION
**Pull request by rdeodoo**
Before this commit:
1. Non Public User would get the main company pricelist returned when calling
   `get_current_pricelist`. As this is saved in an `ir.property`
   (`property_product_pricelist`) that is not website dependant, that pricelist
   would always be returned on every website.
   This is a problem if that pricelist is website limited, you should not be
   able to use it on other websites.
2. Public User would not have a valid pricelist on other website than the
   default one. Indeed, every pricelists are restricted that default website.
   It would result in prices shown as "0" (without currency symbol) on the shop
3. Pricelists set as selectable without a website_id set would not be
   selectable

This commit:
1. We override and check that pricelists returnes by the `ir.property` are
   indeed available on the current website.
2. Remove the website restriction on `Public Pricelist` so other websites have
   at least one pricelist available.
3. `pricelist_ids` on `website` now return all available pricelists including
   the ones that are generic (no website set)

Part of this commit fixes part of #27861
~~Github-25109~~ (Removed from this PR and will be merged in master)
#30743
#30284
#29553

task-1896113
task-1919358
opw-1911789
opw-1935785

### CODE REVIEW

**Comment by rdeodoo** [comment]
Before merge rewrite link to PR (missing hashtag to avoid noise)

File: addons/website_sale/models/product.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def _default_website(self):
     selectable = fields.Boolean(help="Allow the end user to choose this price list")
 
     def clear_cache(self):
-        # website._get_pl() is cached to avoid to recompute at each request the
+        # website._get_pl_partner_order() is cached to avoid to recompute at each request the
```
Comment (JKE-be) [inline addons/website_sale/models/product.py]: merge this commit with another one test commit, since it doesn't fix anything really except maybe a test

File: addons/product/models/product_pricelist.py [hunk 2]
```diff
@@ -310,34 +310,49 @@ def _get_partner_pricelist(self, partner_id, company_id=None):
...
+            Else, it will return the generic property (res_id not set), this one
+                  is created on the company creation.
+            Else, it will return the first available pricelist
+
             :param company_id: if passed, used for looking up properties,
                 instead of current user's company
             :return: a dict {partner_id: pricelist}
         """
-        Partner = self.env['res.partner']
+        Partner = self.env['res.partner'].with_context(active_test=False)
```
Comment (JKE-be) [inline addons/product/models/product_pricelist.py]: comment that you need inactive for public user

is it true ? public user use property too?

File: addons/product/models/product_pricelist.py [hunk 3]
```diff
@@ -310,34 +310,49 @@ def _get_partner_pricelist(self, partner_id, company_id=None):
...
         Property = self.env['ir.property'].with_context(force_company=company_id or self.env.user.company_id.id)
         Pricelist = self.env['product.pricelist']
+        pl_domain = self._get_partner_pricelist_multi_search_domain_hook()
 
-        # retrieve values of property
+        # if no specific property, try to find a fitting pricelist
         result = Property.get_multi('property_product_pricelist', Partner._name, partner_ids)
 
-        remaining_partner_ids = [pid for pid, val in result.items() if not val]
+        remaining_partner_ids = [pid for pid, val in result.items() if not val or not val._get_partner_pricelist_multi_filter_hook()]
```
Comment (JKE-be) [inline addons/product/models/product_pricelist.py]: why partner ?
why multi ?

File: addons/website_sale/models/product.py [hunk 4]
```diff
@@ -17,7 +18,23 @@ class ProductPricelist(models.Model):
     _inherit = "product.pricelist"
 
     def _default_website(self):
-        return self.env['website'].search([], limit=1)
+        """ Find the first company's website, if there is one. """
+        company_id = self.env.user.company_id.id
+        if self.pool._init:
```
Comment (JKE-be) [inline addons/website_sale/models/product.py]: so all test will be faked?

File: addons/website_sale/views/product_views.xml [hunk 5]
```diff
@@ -200,7 +200,7 @@
         <field name="arch" type="xml">
             <field name='currency_id' position='after'>
                 <field name="selectable"/>
-                <field name="website_id" options="{'no_create': True}" string="Allow to use on " placeholder="None website"/>
+                <field name="website_id" options="{'no_create': True}" string="Allow to use on "/>
```
Comment (JKE-be) [inline addons/website_sale/views/product_views.xml]: do we still need label ?

File: addons/product/models/product_pricelist.py [hunk 6]
```diff
@@ -302,6 +302,8 @@ def _price_get_multi(self, pricelist, products_by_qty_by_partner):
             list(pycompat.izip(**products_by_qty_by_partner)))
 
     def _get_partner_pricelist(self, partner_id, company_id=None):
+        # TODO (review): Deprecated? At least not used since https://github.com/odoo/odoo/commit/d39d583b213af02bad12721d9c610def6cb09161
```
Comment (JKE-be) [inline addons/product/models/product_pricelist.py]: to mark as deprecated + nextActivity oto clean in it in master after fwdport

File: addons/website/models/ir_rule.py [hunk 7]
```diff
@@ -7,8 +7,15 @@ class IrRule(models.Model):
...
     @api.model
     def _eval_context(self):
+        is_frontend = self.env['website']._get_request_website()
         res = super(IrRule, self)._eval_context()
-        res['website_id'] = self.env['website'].get_current_website().id
+        # TODO (review): never been used -> res['website_id'] = self.env['website'].get_current_website().id
+
+        # TODO (review): get_current_website(falback=False) still always return a website in backend (if domain set & match)..
+        # This is why is_frontned is needed or in backend we will also see the default website's company items
+        res['website'] = is_frontend and self.env['website'].get_current_website() or self.env['website']
```
Comment (JKE-be) [inline addons/website/models/ir_rule.py]: is_frontend cannot be done in get_current_Website instead? ?

File: addons/website/models/ir_rule.py [hunk 8]
```diff
@@ -7,8 +7,15 @@ class IrRule(models.Model):
...
+        # or this (plus léger mais si un jour on a besoin d'autre chose sur website..)
+        res['website_company_id'] = is_frontend and self.env['website'].get_current_website().company_id or self.env['res.company']
```
Comment (JKE-be) [inline addons/website/models/ir_rule.py]: no res['website'].Company_id seems clean too

File: addons/website_sale/models/product.py (hunk 4 above)
Comment (rdeodoo) [reply addons/website_sale/models/product.py]: Correct me if I am wrong but from what I have tried, `self.pool._init` is always `False` while tests are being runned, both at module install and after install, and with `--test-enable` or `--test-file` (+[1615869d](https://github.com/odoo-dev/odoo/blame/1615869d72c02896c84562c56cb72fcab2b78d6b/addons/mail/models/res_partner.py#L173).)

File: addons/website_sale/models/website.py [hunk 9]
```diff
@@ -75,10 +76,6 @@ def _get_pl_partner_order(self, country_code, show_visible, website_pl, current_
         # This method is cached, must not return records! See also #8795
         return pricelists.ids
 
-    def _get_pl(self, country_code, show_visible, website_pl, current_pl, all_pl):
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/models/website.py]: Done

File: addons/product/models/product_pricelist.py (hunk 6 above)
Comment (rdeodoo) [reply addons/product/models/product_pricelist.py]: Done

File: addons/website/models/ir_rule.py (hunk 8 above)
Comment (rdeodoo) [reply addons/website/models/ir_rule.py]: Done

File: addons/website/models/ir_rule.py (hunk 7 above)
Comment (rdeodoo) [reply addons/website/models/ir_rule.py]: No, `get_current_website` even with `falback=False` still return a website in backend (the one which match the backend URL). I updated the comment to explain that

File: addons/website/models/website.py [hunk 10]
```diff
@@ -447,6 +447,17 @@ def get_url_localized(router, lang):
...
+    @api.model
+    def _get_request_website(self):
+        """ Return the website set on `request` if called in a frontend context
+        (website=True on route).
+        This method can typically be used to check if we are in the frontend.
+
+        This method is easy to mock during python tests to simulate frontend
+        context, rather than mocking every method accessing request.website.
+        """
+        return request and hasattr(request, 'website') and request.website or False
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/models/website.py]: Done.
It comes with a cost, as now it is not a Class method anymore, we need to import it to use it, but we can't import it directly or the mock won't work as import in files will 'override' it:
It became
```python
from odoo.addons.website.models import ir_http
ir_http.get_request_website()
```
Instead of
```python
self.env['website'].get_request_website()
```
See [this blog](https://blog.swdev.ed.ac.uk/2017/04/06/how-you-import-impacts-how-you-mock/). 
I am ok with that, let me know if something bugs you

File: addons/website_sale/models/res_partner.py [hunk 11]
```diff
@@ -19,7 +19,6 @@ def _compute_last_website_so_id(self):
             if website and not is_public:
                 partner.last_website_so_id = SaleOrder.search([
                     ('partner_id', '=', partner.id),
-                    ('team_id.team_type', '=', 'website'),
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/models/res_partner.py]: About that, do you have an easy/practical way to [keep commit order](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/) on github.com after an interactive rebase?
Right now, I need to `git rebase -i HEAD~9` and replace `pick` by `r` on every commit to rewrite their message to have a correct date order and so a correct github.com order, but it is kinda annoying..

[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/models/res_partner.py]: Actually found something, in case you need it:
```bash
[alias]
    reorder = "!f() { git rebase \"$1\" --ignore-date -x 'git commit --amend -C HEAD --date=\"$(date -R)\" && sleep 0.10'; }; f"
```
Then just `git reorder HEAD~8`

File: addons/product/models/product_pricelist.py (hunk 2 above)
Comment (rdeodoo) [reply addons/product/models/product_pricelist.py]: Yes of course (from what I understood from current code implementation).
The website public user is inactive.
Still, we need to retrieve its `property_product_pricelist` as [it will be passed](https://github.com/odoo-dev/odoo/blame/905a4a4be435011668f6994bb2d0e4d16ff2e5aa/addons/website_sale/models/website.py#L127) to the `_get_pl_partner_order()` method as the [default website pricelist](https://github.com/odoo-dev/odoo/blame/905a4a4be435011668f6994bb2d0e4d16ff2e5aa/addons/website_sale/models/website.py#L55) (`website_pl`) parameter.
I already wrote a test for that and explained why, see [this test](https://github.com/odoo-dev/odoo/blame/905a4a4be435011668f6994bb2d0e4d16ff2e5aa/addons/website_sale/tests/test_website_sale_pricelist.py#L232).

**Comment by rdeodoo** [comment]
@JKE-be Thanks a lot for your review, I processed everything and replied to your questions.

@nim-odoo As discussed, your input is more than welcome.
This PR first goal was to fix multi-website pricelists but it actually fixes more than that, mainly multi-company behaviors too (without multi-website) and `property_product_pricelist` compute (public user etc).
The only thing I am really not sure about is the last `if` of `_get_pl_partner_order()`, you will see a comment anyway.

Basically, I know what I fixed (and tested a lot of it) but I am not sure of what I could have broke, and that's what I am worried about.

Note that I tried to split in separate commits for clearness purpose, first 4 commits don't do much beside cleaning some stuff.

Thank you!

File: addons/website_sale/models/website.py [hunk 12]
```diff
@@ -59,7 +59,7 @@ def _get_pl_partner_order(self, country_code, show_visible, website_pl, current_
         pricelists = self.env['product.pricelist']
         if country_code:
             for cgroup in self.env['res.country.group'].search([('country_ids.code', '=', country_code)]):
-                for group_pricelists in cgroup.pricelist_ids:
+                for group_pricelists in cgroup.pricelist_ids.filtered(lambda pl: pl.website_id.id in (False, self.id)):
```
Comment (nim-odoo) [inline addons/website_sale/models/website.py]: Can't you add this in the `search` instead of filtering afterwards?

File: addons/website_sale/__init__.py [hunk 13]
```diff
@@ -1,7 +1,17 @@
...
 from . import wizard
 from . import report
+
+
+def uninstall_hook(cr, registry):
+    ''' Need to reenable the `product` pricelist multi-company rule that were
+        disabled to be 'overriden' for multi-website purpose
+    '''
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    multi_company_rules = env.ref('product.product_pricelist_comp_rule', raise_if_not_found=False)
```
Comment (nim-odoo) [inline addons/website_sale/__init__.py]: This will fail if one the the rules can't be found:
```
In [3]: multi_company_rules = self.env.ref('product.product_pricelist_comp_rule', raise_if_not_found=False)

In [4]: multi_company_rules
Out[4]: ir.rule(32,)

In [5]: multi_company_rules += self.env.ref('product.product_pricelist_comp_ruleXXXX', raise_if_not_found=False)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-c4fe14dd8fc2> in <module>()
----> 1 multi_company_rules += self.env.ref('product.product_pricelist_comp_ruleXXXX', raise_if_not_found=False)

/home/nim/src/odoo/odoo/models.py in __add__(self, other)
   4674     def __add__(self, other):
   4675         """ Return the concatenation of two recordsets. """
-> 4676         return self.concat(other)
   4677 
   4678     def concat(self, *args):

/home/nim/src/odoo/odoo/models.py in concat(self, *args)
   4683         for arg in args:
   4684             if not (isinstance(arg, BaseModel) and arg._name == self._name):
-> 4685                 raise TypeError("Mixing apples and oranges: %s.concat(%s)" % (self, arg))
   4686             ids.extend(arg._ids)
   4687         return self.browse(ids)

TypeError: Mixing apples and oranges: ir.rule(32,).concat(None)

```

File: addons/website_sale/models/website.py [hunk 14]
```diff
@@ -60,35 +60,53 @@ def _get_pl_partner_order(self, country_code, show_visible, website_pl, current_
...
-            pricelists |= all_pl.filtered(lambda pl: not show_visible or pl.selectable or pl.id in (current_pl, order_pl))
-        if not show_visible and not country_code:
-            pricelists |= all_pl.filtered(lambda pl: pl.sudo().code)
+        # If the user is logged in and either he has no country pricelists or either
+        # his partner pricelist is not the website pricelist, then we use his partner pricelist
+        if not is_public and (not pricelists or partner_pl != website_pl):
+            pricelists |= pricelists.browse(partner_pl)
+
+        if is_public and not pricelists:  # no pricelist for this country, or no GeoIP
+            # TODO (review) if is_public was implicit as you cant go in this is if you are not public given the previous if
```
Comment (nim-odoo) [inline addons/website_sale/models/website.py]: What?

File: addons/website_sale/models/website.py [hunk 15]
```diff
@@ -101,18 +119,21 @@ def get_pricelist_available(self, show_visible=False):
...
         pricelists = website._get_pl_partner_order(isocountry, show_visible,
                                                    website.user_id.sudo().partner_id.property_product_pricelist.id,
-                                                   request and request.session.get('website_sale_current_pl') or None,
+                                                   req and req.session.get('website_sale_current_pl') or None,
                                                    website.pricelist_ids,
                                                    partner_pl=partner_pl and partner_pl.id or None,
                                                    order_pl=last_order_pl and last_order_pl.id or None)
         return self.env['product.pricelist'].browse(pricelists)
 
+    def get_pricelist_available(self, show_visible=False):
```
Comment (nim-odoo) [inline addons/website_sale/models/website.py]: To remove in master?

Comment (rdeodoo) [reply addons/website_sale/models/website.py]: I encapsulated the `get_pricelist_available()` in `_get_pricelist_available()` to be able to easily mock the method.
So Odoo's code is still calling `get_pricelist_available()`, it should not be removed.

See this code in `test_website_sale_pricelist.py` that is in `[FIX] website_sale: fix _get_pl_partner_order` commit of this PR:
```python
class DotDict(dict):
    """dot.notation access to dictionary attributes"""
    def __getattr__(*args):
        val = dict.get(*args)
        return DotDict(val) if type(val) is dict else val

def get_pricelist_available(self, show_visible=False, website_id=1, country_code=None, website_sale_current_pl=None):
    request = DotDict({
        'website': self.env['website'].browse(website_id),
        'session': {
            'geoip': {
                'country_code': country_code,
            },
            'website_sale_current_pl': website_sale_current_pl,
        },
    })
    return self.env['website']._get_pricelist_available(request, show_visible)
```

I know it is a bit weird but to test `get_pricelist_available()` from python, it is the only way really to simulate a frontend context, as:
1. We don't want to test this method in tour for performance reason
2. We want to write lot of tests on that method as it is the core method of `website_sale`
3. We can't use `request` in tests as it is unbound at that time and `get_pricelist_available()` is trying to read `request` attributes (see [here](https://github.com/odoo/odoo/blame/d12c993a28b5971dc5bf7dbb88e4b7c606f592b5/addons/website_sale/models/website.py#L89), [here](https://github.com/odoo/odoo/blame/d12c993a28b5971dc5bf7dbb88e4b7c606f592b5/addons/website_sale/models/website.py#L96) and [here](https://github.com/odoo/odoo/blame/d12c993a28b5971dc5bf7dbb88e4b7c606f592b5/addons/website_sale/models/website.py#L102).)

It is the cleanest way I came up with to test `website_sale` behaviors regarding pricelists.
That way, Odoo's code still use `get_pricelist_available()` that call `_get_pricelist_available()` and pass `request` while tests call directly `_get_pricelist_available()` and give `DotDict` class as request parameter (that mock perfectly the regular request).

File: addons/website_sale/models/website.py (hunk 14 above)
Comment (rdeodoo) [reply addons/website_sale/models/website.py]: Knowing that `partner_pl` is the `property_product_pricelist` of the partner, that field is now already 'website compliant' with this PR (in a frontend context).

The previous if is:
```python
if not is_public and (not pricelists or partner_pl != website_pl):`
    pricelists |= pricelists.browse(partner_pl)
```

So after that condition,
```python
if not pricelists:
```
Is the same as
```python
if is_public and not pricelists:
```
As the previous if, if you are not public and you don't have a pricelist will add `partner_pl` in the `pricelists` variable. So if you don't have a pricelist at that time, you are by deduction public.

Unless I am missing something :yum:

File: addons/website_sale/__init__.py (hunk 13 above)
Comment (rdeodoo) [reply addons/website_sale/__init__.py]: Well spotted, I added `or env['ir.rule']` in case of not found, if the user deleted himself the records.

File: addons/website_sale/models/website.py (hunk 12 above)
Comment (rdeodoo) [reply addons/website_sale/models/website.py]: I don't think so. By searching `country.group` for the given `country_code` and having a pricelist 'website available' , I would still have to refilter the pricelist_ids afterwards as a `res.country.group` might have at least one pricelist 'website available' without especially having all its pricelists 'website available'.

Plus, I would have to mimick the `_get_website_pricelists_domain()` method (that is a domain supposed to be used on `product.pricelist` directly.

Here, to add the domain in the search, I would have to copy the 'accessible on website' condition that I would like to keep encapsulated only in `_get_website_pricelists_domain()` (and its filter equivalent `_is_available_on_website()`).

Currently, it is:
```python
if country_code:
    for cgroup in self.env['res.country.group'].search([('country_ids.code', '=', country_code)]):
        pricelists |= cgroup.pricelist_ids.filtered(lambda pl: pl._is_available_on_website(self.id) and _check_show_visible(pl))
```
If I add the domain in search, it would become:
```python
if country_code:
    domain = [
        ('country_ids.code', '=', country_code),
        '|', ('pricelist_ids.website_id', '=', website_id),
        '&', ('website_id', '=', False),
        '|', ('selectable', '=', True), ('code', '!=', False),
    ]
    for cgroup in self.env['res.country.group'].search([('country_ids.code', '=', country_code)]):
        pricelists |= cgroup.pricelist_ids.filtered(lambda pl: pl._is_available_on_website(self.id) and _check_show_visible(pl))
```

But maybe it is worth speeding a little bit the search with the domain even if we need to refilter in python afterward?

File: addons/website_sale/models/website.py (hunk 15 above)
Comment (nim-odoo) [reply addons/website_sale/models/website.py]: ok

**Comment by rdeodoo** [comment]
@nim-odoo Thanks for your review.

If there is some corner case you can think about that had to be fixed multiple times or that always been subject to misbehavior or that you know works, are important but not tested, let me know.
Other than that, I am pretty confident in behaviors I fixed/implemented, multi-company/multi-website needed that relifting to work.
But I am still worried about any side effects of potential specific flows I am not aware of (backend?). I know pricelists have always been subject to debate/fixes.

Anyway I did my best :yum:, I will finish @JKE-be 's review and then we will see

**Comment by nim-odoo** [comment]
I'm ok with the change, no problem. I'm just like 99 % sure strike-through prices are gonna be broken in a way of another, but we are used to that :smile: 
We'll try to include tests if we need to fix corner cases later.

File: addons/website_sale/models/product.py [hunk 16]
```diff
@@ -17,7 +19,23 @@ class ProductPricelist(models.Model):
...
+            # During l10n modules install, it will change the company currency
+            # that will create a default pricelist for that currency.
+            # We should not search for a website in that case to avoid
+            # `_check_websites_in_company` constraint.
+            return self.env['website']
```
Comment (JKE-be) [inline addons/website_sale/models/product.py]: no more needed if we move constraint in master

but always not understand this case.
It is louchhhh to make special case, to bypass a new constraint...

File: addons/website_sale/models/product.py [hunk 17]
```diff
@@ -32,6 +50,8 @@ def clear_cache(self):
 
     @api.model
     def create(self, data):
+        if data.get('company_id') and not data.get('website_id'):
```
Comment (JKE-be) [inline addons/website_sale/models/product.py]: what is the goal of this?

File: addons/website_sale/models/website.py [hunk 18]
```diff
@@ -60,35 +60,53 @@ def _get_pl_partner_order(self, country_code, show_visible, website_pl, current_
...
+            # TODO (review) this if condition as always been weirdoooo
+            # Introduced with https://github.com/odoo/odoo/commit/60300fc287928358d31ae5faff1b7f9953c57d18
+            # to allow logged in user to be able to use promo code but it should actually have been this directly, not just filtered on code
+            # Then fixed with https://github.com/odoo/odoo/commit/e11908d429435091828f224ee0751b90de8c99aa
+            # to allow user to see selectable pricelists but the fix makes no sense as it still only filter on code..
+            pricelists |= all_pl.filtered(lambda pl: _check_show_visible(pl))
 
         # This method is cached, must not return records! See also #8795
-        return pricelists.filtered(lambda pl: pl._is_available_on_website(self.id)).ids
+        return pricelists.ids
```
Comment (JKE-be) [inline addons/website_sale/models/website.py]: can we simplify with:


```python
if country_code:
    for cgroup in self.env['res.country.group'].search([('country_ids.code', '=', country_code)]):
        pricelists |= cgroup.pricelist_ids.filtered(lambda pl: pl._is_available_on_website(self.id) and _check_show_visible(pl))
if not country_code or not pricelists:
    pricelists |= all_pl.filtered(lambda pl: _check_show_visible(pl))


is_public = self.user_id.id == self.env.user.id
if not is_public:
    pricelists |= pricelists.browse(partner_pl)
```

 + need

```python
# add current pl (e.g. if promo pl, country_code not set on pricelist (ignored btw), but set in method)
pricelists |= pricelists.browse(current_pl)
```

**Comment by rdeodoo** [comment]
About perfs, I did some tests, see this [spreadsheet](https://docs.google.com/spreadsheets/d/1IZ7HYY1Vo3QvzCYLS9HKLtLC0eH9CET25cnLuK0RE8I/edit?usp=sharing) for results but basically:

- For **public user**, went from **597 to 612** requests for a total of **0.231s vs 0.236s** after my fix (based on 200 tests).

- For **logged in user**, went from **653 to 662** requests for a total of **0.294s vs 0.304s** after my fix (based on only 20 tests, I could do more if needed..).

File: addons/website_sale/models/res_partner.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/models/res_partner.py]: Done at https://github.com/odoo/odoo/pull/31669

File: addons/website_sale/static/src/js/website_sale_tour_shop_dynamic_variants.js [hunk 19]
```diff
@@ -26,6 +26,7 @@ tour.register('tour_shop_dynamic_variants', {
     },
     {
         content: "click add to cart",
+        extra_trigger: 'body:has(input[type="hidden"][name="product_id"][value=0])',
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/static/src/js/website_sale_tour_shop_dynamic_variants.js]: Done at https://github.com/odoo/odoo/pull/31670

File: addons/website/models/ir_http.py [hunk 20]
```diff
@@ -41,6 +41,23 @@ def sitemap_qs2dom(qs, route, field='name'):
...
+    This method is easy to mock during python tests to simulate frontend
+    context, rather than mocking every method accessing request.website.
+
+    Don't import directly the method or it won't be mocked during tests, do:
+    ```
+    from odoo.addons.website.models import ir_http
+    my_var = ir_http.get_request_website()
+    ```
+    """
+    return request and getattr(request, 'website', False) or False
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/models/ir_http.py]: I agree but since it is not a model method I can't easily return an empty recordset.
And we don't want it to be a website model method as we want to keep it on http.
So as discussed, for now we will keep False as possible return value.

File: addons/website_sale/views/product_views.xml (hunk 5 above)
Comment (rdeodoo) [reply addons/website_sale/views/product_views.xml]: Alright, done, it will use the fields `string` as label, I added the uppercase on Website.

File: addons/website/models/ir_rule.py [hunk 21]
```diff
@@ -8,7 +9,14 @@ class IrRule(models.Model):
...
         res = super(IrRule, self)._eval_context()
-        res['website_id'] = self.env['website'].get_current_website().id
+
+        # We need is_frontend to avoid showing website's company items in backend
+        # (that could be different than current company). We can't use
+        # `get_current_website(falback=False)` as it could also return a website
+        # in backend (if domain set & match)..
+        is_frontend = ir_http.get_request_website()
+        Website = self.env['website']
+        res['website'] = is_frontend and Website.get_current_website() or Website
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/models/ir_rule.py]: As discussed, we will keep returning False as it is not a model method, no access to self, and request might be unbound. So it is hard to return an empty record set

File: addons/product/models/product_pricelist.py (hunk 2 above)
Comment (rdeodoo) [reply addons/product/models/product_pricelist.py]: As discussed, it was really a unwanted/unnoticed side effect of https://github.com/odoo-dev/odoo/commit/d39d583b213af02bad12721d9c610def6cb09161
It would do
```python
p = Partner.browse(partner_id)
```
And now it will do
```python
for pid in Partner.search(group['__domain']).ids:
```
Browse would return inactive while search dont.
See last message for explanation about website public user:

> Still, we need to retrieve its property_product_pricelist as it will be passed to the _get_pl_partner_order() method as the default website pricelist (website_pl) parameter.
> I already wrote a test for that and explained why, see this test.

File: addons/website_sale/models/product.py (hunk 16 above)
Comment (rdeodoo) [reply addons/website_sale/models/product.py]: Constraint about misconfiguration on multi-website/multi-company and its related `if` on `_default_website` have been removed from this PR and will be added back in master in another PR.

File: addons/website_sale/models/product.py [hunk 22]
```diff
@@ -48,6 +68,72 @@ def unlink(self):
...
+
+    def _get_partner_pricelist_multi_filter_hook(self):
+        res = super(ProductPricelist, self)._get_partner_pricelist_multi_filter_hook()
+        website = ir_http.get_request_website()
+        if website:
+            res = res.filtered(lambda pl: pl._is_available_on_website(website.id))
+        return res
+
+    @api.onchange('company_id')
+    def _onchange_company_id(self):
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/models/product.py]: Constraint about misconfiguration on multi-website/multi-company and its related if on _default_website have been removed from this PR and will be added back in master in another PR.

File: addons/website_sale/models/product.py [hunk 23]
```diff
@@ -48,6 +68,72 @@ def unlink(self):
...
+        ''' Show only the company's website '''
+        domain = self.company_id and [('company_id', '=', self.company_id.id)] or []
+        return {'domain': {'website_id': domain}}
+
+    @api.constrains('company_id', 'website_id')
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/models/product.py]: Constraint about misconfiguration on multi-website/multi-company and its related if on _default_website have been removed from this PR and will be added back in master in another PR.

File: addons/website_sale/models/product.py [hunk 24]
```diff
@@ -17,21 +19,39 @@ class ProductPricelist(models.Model):
...
+            # We should not search for a website in that case to avoid
+            # `_check_websites_in_company` constraint.
+            return self.env['website']
+
+        if self._context.get('default_company_id'):
+            # Also, when changing a company currency, it will create a pricelist for
+            # that currency, we should not find a website for the user current company
+            # but one for the company being edited.
+            company_id = self._context.get('default_company_id')
+
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/models/product.py]: Constraint about misconfiguration on multi-website/multi-company and its related if on _default_website have been removed from this PR and will be added back in master in another PR.

File: addons/website_sale/models/product.py (hunk 1 above)
Comment (rdeodoo) [reply addons/website_sale/models/product.py]: As discussed, I will let this commit alone on top of the PR to avoid rebase nightmares.

File: addons/website_sale/models/website.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/models/website.py]: Indeed, done.

File: addons/website_sale/models/website.py [hunk 25]
```diff
@@ -54,57 +60,80 @@ def _get_pl_partner_order(self, country_code, show_visible, website_pl, current_
...
-    def get_pricelist_available(self, show_visible=False):
-
+    def _get_pricelist_available(self, req, show_visible=False):
         """ Return the list of pricelists that can be used on website for the current user.
         Country restrictions will be detected with GeoIP (if installed).
         :param bool show_visible: if True, we don't display pricelist where selectable is False (Eg: Code promo)
         :returns: pricelist recordset
         """
-        website = request and hasattr(request, 'website') and request.website or None
+        website = ir_http.get_request_website()
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_sale/models/website.py]: Given the commit order, yes:

**At first**, first commit of the PR creates the `get_request_website` and replace occurences of
```python
request and hasattr(request, 'website') and request.website or None
```
by
```python
ir_http.get_request_website()
```

**In second**, the last commit of this PR encapsulate the code of `get_pricelist_available` in private method `_get_pricelist_available` so `get_pricelist_available` can be mocked easily and fully tests in python the pricelists behaviors.

So when `website = ir_http.get_request_website()` is introduced, `req` and `_get_pricelist_available` does not exists.

Let me know if this is a problem, I will change last commit so it becomes:
```python
def get_pricelist_available(self, show_visible=False):
    website = ir_http.get_request_website()
    return self._get_pricelist_available(request, website, show_visible)

def _get_pricelist_available(self, req, website, show_visible=False):
    ..
```
That way `ir_http.get_request_website()` and `req` won't be in the same method (`_get_pricelist_available`) but not sure it is cleaner..

File: addons/website_sale/models/website.py (hunk 18 above)
Comment (rdeodoo) [reply addons/website_sale/models/website.py]: Done.
About
```python
# add current pl (e.g. if promo pl, country_code not set on pricelist (ignored btw), but set in method)
pricelists |= pricelists.browse(current_pl)
```
I added a test `test_get_pricelist_available_geoip4` that test that current_pl is added and it works without that line.

So I am probably not understanding your flow (public user? how can he set its current_pl into session if not shown in the first place?) let me know.

**Comment by rdeodoo** [comment]
Every review has been processed.
Code has been cleaned, especially from TODO's and review comments.

**Comment by nim-odoo** [comment]
As said previously, nothing was striking me based on the diff, but I know pricelists and website have never been a good combination. Just one last remark: can you make sure it doesn't cause problems if you don't upgrade the module (only update the Python sources)?

File: addons/website_sale/models/product.py [hunk 26]
```diff
@@ -91,6 +91,17 @@ def _get_website_pricelists_domain(self, website_id):
...
 
+    def _get_partner_pricelist_multi(self, partner_ids, company_id=None):
+        ''' If `property_product_pricelist` is read from website, we should use
+            the website's company and not the user's one.
+            Passing a `company_id` to super will avoid using the current user's
+            company.
+        '''
+        website = ir_http.get_request_website()
+        if not company_id and website:
+            company_id = website.company_id.id
```
Comment (rdeodoo) [inline addons/website_sale/models/product.py]: @nim-odoo From my tests no error on python reload with this PR without an update.
Except this case:
What might happen is that user won't get the new `ir.rule` on company pricelists to be able to read pricelists from website's company in addition to his company ones.

In that case, a specific setup might lead to an access error: A partner on company 1 which has a pricelist from company 1 specifically set on him. Then if that partner visit a website which is from another company (not allowed for that partner) it will raise an error as no access to company 2 without the new ir.rules.

We could add a condition here, a try/except to be sure the user can access the pricelists from the website's company.
If the try fails, it means the module was not updated and we can't use the website's company.

Something like
```python
if not company_id and website:
    # Compatibility check, `company_id = website.company_id.id` has to be
    # conditionned. Indeed new ir.rule to access pricelists from website's
    # company might not exists (reload python without module update).
    # Remove me in saas-12.1
    try:
        website.company_id.check_access_rule('read')
        company_id = website.company_id.id
    except:
        pass
```

Note that currently you already have access error on standard code when setuping multi-company/multi-website environment, but mostly 403 and not 500.

Comment (nim-odoo) [reply addons/website_sale/models/product.py]: Hum, I don't get it. The new rules apply to product.pricelist and product.pricelist.item, so why would it fail on website.company_id.id? Anyway that's better to always foresee a fallback in these cases. It's also better to test except AccessError

Comment (rdeodoo) [reply addons/website_sale/models/product.py]: Actually reading the `id` `website.company_id.id` will not raise any traceback even if the user can't access the company. I could have tried to read `website.company_id.name` instead of `website.company_id.check_access_rule('read')`.

But if we pass `website.company_id.id` as company and the user does not have the new `ir.rule`, it will throw an error since:
1. We will try to read `property_product_pricelist` which is a compute
2. that is calling `_get_partner_pricelist_multi` which does
```python
Property = self.env['ir.property'].with_context(force_company=company_id or self.env.user.company_id.id)
result = Property.get_multi('property_product_pricelist', Partner._name, partner_ids)
```

Then it would return the pricelists from the website company, and if the new `ir.rule` are not there, it will raise an error if the website's company is not the user's company.

Comment (rdeodoo) [reply addons/website_sale/models/product.py]: Indeed. I correctly fixed the compatibility issue.
The case is still:

> A partner on company 1 which has a pricelist from company 1 specifically set on him. Then if that partner visit a website which is from another company (not allowed for that partner) it will raise an error as no access to company 2 pricelists without the new ir.rules.

I pushed a commit that I will amend if it looks ok to you. You can directly check it.
Basically, I try to read the website's company's pricelists, if it raise an error, it means the DB did not update `website_sale` and does not have the new `ir.rule` allowing website's company pricelists to be accessed. In that case, I just fallback on old implementation.

Comment (nim-odoo) [reply addons/website_sale/models/product.py]: ok :smile:

**Comment by JKE-be** [comment]
@robodoo r+ rebase-ff

**Comment by rdeodoo** [comment]
@JKE-be @nim-odoo Thanks a million times for the review(s).
